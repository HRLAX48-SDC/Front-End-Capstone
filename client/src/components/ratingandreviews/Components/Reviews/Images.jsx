import React, { useState, useEffect} from 'react';
import { Card, Button, Modal, Image, Row, ModalBody } from 'react-bootstrap'



function Images ({ image }) {
    const[imageLarge, setImageLarge]= useState(false)

    return (
        <>
        <Image className="review-photos" thumbnail src={image.url} onClick={() => setImageLarge(true) }/>
        <Modal show={imageLarge} centered size="large" onHide={() => setImageLarge(false)}>
            <ModalBody>
                <Image src={image.url} thumbnail/>
            </ModalBody>
        </Modal>
        </>
    )
}

export default Images;
