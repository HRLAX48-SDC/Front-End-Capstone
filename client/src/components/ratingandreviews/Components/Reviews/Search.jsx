import React, {useState, useEfect} from 'react';
import { Form, InputGroup } from 'react-bootstrap';

const Search = ({search, setSearch}) => {
  return (
    <Form>
        <InputGroup>
        <InputGroup.Text>
        <i className="fas fa-search"></i>
        </InputGroup.Text>
        
        <Form.Control
            type="text"
            placeholder="Search reviews..."
            value={search} 
            onChange={(e) => setSearch(e.target.value)}
            />
        </InputGroup>
    </Form>
  )
}
export default Search;