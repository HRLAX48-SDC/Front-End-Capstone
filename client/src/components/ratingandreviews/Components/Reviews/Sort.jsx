import React, { useState, useEffect} from 'react';
import Dropdown from 'react-bootstrap/Dropdown'
import DropdownButton from 'react-bootstrap/DropdownButton'


function Sort (props) {

  //make the tile of the button change to what item was clicked on
  const[title, setTitle]= useState('Relevance');

  const changeTitle = (e) =>{
    setTitle(e.target.name)
  } 

    return (
      <DropdownButton id="dropdown-basic-button" title={title} variant={'secondary'} >
        <Dropdown.Item href="#/action-1" name="Helpful" onClick={changeTitle}>Helpful</Dropdown.Item>
        <Dropdown.Item href="#/action-2" name="Newest"onClick={changeTitle}>Newest</Dropdown.Item>
        <Dropdown.Item href="#/action-3" name="Relevance" onClick={changeTitle}>Relevance</Dropdown.Item>
      </DropdownButton>
    )
}
export default Sort;

