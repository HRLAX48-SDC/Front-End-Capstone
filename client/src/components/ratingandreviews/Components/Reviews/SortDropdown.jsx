import React, { useState, useEffect} from 'react';
import Dropdown from 'react-bootstrap/Dropdown'
import DropdownButton from 'react-bootstrap/DropdownButton'


function SortDropdown (props) {

  //make the tile of the button change to what item was clicked on
  const[title, setTitle]= useState('Relevance');
  
  useEffect(() => {
    props.sortBy(title)
  }, [title])
  
  const changeTitle = (e) =>{
    setTitle(e.target.name)
    props.sortBy(title)
  } 
  // const sortBy = () => {
  //   props.sortBy(title)
  // }


    return (
      <DropdownButton id="dropdown-basic-button" title={title} variant={'secondary'} >
        <Dropdown.Item href="#/action-1" name="Relevant" onClick={changeTitle}>Relevant</Dropdown.Item>
        <Dropdown.Item href="#/action-2" name="Newest" onClick={changeTitle}>Newest</Dropdown.Item>
        <Dropdown.Item href="#/action-3" name="Helpful" onClick={changeTitle}>Helpful</Dropdown.Item>
      </DropdownButton>
    )
}
export default SortDropdown;

