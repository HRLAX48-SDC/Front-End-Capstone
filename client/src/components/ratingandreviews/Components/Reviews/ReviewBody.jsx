import React, { useState, useEffect} from 'react';
import { Card, Button, Modal, Image, Row } from 'react-bootstrap'
import Images from './Images.jsx';

function ReviewBody (props) {
  const slicedBody = props.review.body.slice(0, 250);
  const reviewPhotos = props.review.photos;
  const [body, setBody] = useState(slicedBody);

  
  const showMore = () => {
    setBody(props.review.body)
  }


  return (
    <Card>
      <Card.Body>
        <Card.Text>
          {body}
        </Card.Text>
          {props.review.body.length > 250 &&
          <Button variant="outline-secondary"onClick={() => showMore()}>Show More</Button>
          }
          <Row>
            {reviewPhotos.length > 0 &&
            <div>
              {reviewPhotos.map((image, index) => 
                  <Images image={image} key={index}/>
              )}
            </div>
            }
          </Row>
      </Card.Body>
    </Card>
  )
}
export default ReviewBody;


//optional extras:

//default to first 250 chars
  //show more button should only appear when there are more than 250 chars

//display a "..." when review is trimmed
  {/* {props.review.body.length > 250 ? <>{body}...</>
          : <>{body}</>} */}