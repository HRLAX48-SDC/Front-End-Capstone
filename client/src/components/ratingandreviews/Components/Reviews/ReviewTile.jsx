import React, { useState, useEffect} from 'react';
import moment from 'moment';
import ReviewBody from './ReviewBody.jsx';
import { Card, Stack } from 'react-bootstrap';
import API_KEY from '../../../../config/config.js';
import axios from 'axios';


function ReviewTile (props) {

  const [reviewHelpfulNum, setReviewHelpfulNum] = useState(() => props.review.helpfulness);
  const [helpfulSelection, setHelpfulSelection] = useState(false);
 
  
  const selectHelpful = () =>{
    if (!helpfulSelection) {
    axios.put(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-lax/reviews/${props.review.review_id}/helpful`, 
    {helpfulness: reviewHelpfulNum + 1},
    { headers: { 'Authorization': `${API_KEY}` } })
    .then(res => {
      setReviewHelpfulNum(reviewHelpfulNum + 1);
      setHelpfulSelection(true);
    })
    .catch(err => console.log(err));
    }
  }

  const selectUnHelpful = () =>{
    if (!helpfulSelection) {
    axios.put(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-lax/reviews/${props.review.review_id}/helpful`, 
    {helpfulness: reviewHelpfulNum - 1},
    { headers: { 'Authorization': `${API_KEY}` } })
    .then(res => {
      setReviewHelpfulNum(reviewHelpfulNum - 1);
      setHelpfulSelection(true);
    })
    .catch(err => console.log(err));
    }
  }

  //recommended
  //if the review is recommended display "this review is recommended"

  //make put request to report reviews...


  return (
    <>
      <Card>
        <Card.Header>
          <Stack direction="horizontal" gap={3}>
            <div>*star rating*</div>
            <div className="ms-auto">{props.review.reviewer_name}</div>
            <div>{moment(props.review.date).format('MMMM Do, YYYY')}</div>
          </Stack>
        </Card.Header>
        <Card.Body>
          <Card.Title>{props.review.title}</Card.Title>
          <Card.Text>
          {props.review.summary}
          </Card.Text>
          <ReviewBody review={props.review}/>
          <Stack direction="horizontal" gap={3}>
            <div>Helpful?</div>
            <div onClick={() => selectHelpful()}>Yes</div>
            <div onClick={() => selectUnHelpful()}>No</div>
            <div>({reviewHelpfulNum})</div>
            <div>*report*</div>
           </Stack>
        </Card.Body>
        
      </Card>
    </>
  )

}
export default ReviewTile;

