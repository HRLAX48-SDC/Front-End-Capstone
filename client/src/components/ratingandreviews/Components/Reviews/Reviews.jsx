import React, { useState, useEffect} from 'react';
import axios from 'axios';
import API_KEY from '../../../../config/config.js'
// import Card from 'react-bootstrap/Card';
// import Button from 'react-bootstrap/Button';
import { Button, Card, Stack } from 'react-bootstrap';
import ReviewTile from './ReviewTile.jsx';
import Sort from './Sort.jsx';


function Reviews (props) {

    // const [moreReviews, setMoreReviews] = useState(false); //display button if more than two reviews
    const [reviewList, setReviewList] = useState([]); //get initial reviewArray
    const [reviewNum, setReviewNum] = useState(2); //display 2 tiles initially and more when moreReviews is true
    
    
    const getReviews =() => { //currently gets one by id
    axios.get(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-lax/reviews/?product_id=42366`, {
        headers: {
        'Authorization': `${API_KEY}`
        },
    })
    .then((response) => {
        setReviewList(response.data.results)
    })
    .catch((err) => {
        console.log(err);
    })
    }

    useEffect(() => {
    getReviews()
    }, []);

    const slicedReviews = reviewList.slice(0, reviewNum);
    const loadMore = () => {
        setReviewNum(reviewNum + 2);
    }


    return (
        <Card>
        <Card.Header>{reviewList.length} reviews, sorted by: <Sort /></Card.Header>
        <Card.Body>
        {reviewList ? 
            <Stack gap={3}>
                {slicedReviews.map(review => (
                    <ReviewTile key={review.review_id} review={review}/>
                ))}
            </Stack>
            : 
            <div className="spinner-border" role="status">
                <span className="visually-hidden">Loading...</span>
            </div>
            }
            <>
              <Button variant="outline-secondary">Submit Review</Button>{' '}
              {reviewList.length > 2 &&
              <Button variant="outline-secondary" onClick={() => loadMore()}>More Reviews</Button>
              }
            </>
        </Card.Body>
      </Card>
    )
};
export default Reviews;


  //if tileNum > 2
        //show more review button
    // const getReviewNum = () => {
    //     if (props.reviews.count) {
    //         setTileNum(props.reviews.count);
    //     }
    // }



// <div>
//     <button>Sort Dropdown</button>
//     <button>Submit Review</button>
//     {props.reviews.results ? 
//     <div>
//         {props.reviews.results.map(tile => (
//             <Tile key={tile.review_id} review={tile}/>
//         ))}
//     </div>
//     : 
//     <div className="spinner-border" role="status">
//         <span className="visually-hidden">Loading...</span>
//     </div>
//     }
//     <button>More Reviews</button>
// </div>



// class List extends React.Component {
//   constructor(props){
//     super(props);
    
//     this.state = {
//       reviews: {}
//     }
    
//     }
//     componentDidMount(){
//         // this.props.getReviews()
        
//     }
    

//   render() {
    
//       return (
//         <div>
//             <button>Sort Dropdown</button>
//             <button>Submit Review</button>
//             <div>
//                 {
//                     // props.reviews.results.map(tile => (
//                 //     <Tile key={tile.review_id} review={tile}/>
//                 // ))
//                 }
//             </div>
//             <button>More Reviews</button>
//         </div>
//       )
//   }
// }