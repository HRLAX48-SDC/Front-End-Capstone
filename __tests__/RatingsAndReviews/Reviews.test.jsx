import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Reviews from '../../client/src/components/ratingandreviews/Components/Reviews/Reviews.jsx'

Enzyme.configure({adapter: new Adapter()});


describe('<Reviews />', () => {
    const reviews = [{
        body: "Wow this product is amazing, 100/10 recommend. I own 1000 of these muahahaha. ",
        date: "2021-11-10T00:00:00.000Z",
        helpfulness: 6,
        photos: [],
        rating: 5,
        recommend: true,
        response: null,
        review_id: 1094674,
        reviewer_name: "BananaBoy123",
        summary: "Is this really the best purchase ever?"
    }] 
    const metaData = {
        characteristics :{
            Comfort: {id: 142034, value: "3.1627906976744186"},
            Fit: {id: 142032, value: "2.9021739130434783"},
            Length: {id: 142033, value: "2.8045977011494253"},
            Quality: {id: 142035, value: "3.1976744186046512"}
        },
        product_id: "42366",
        ratings: {1: "23", 2: "8", 3: "50", 4: "36", 5: "47"},
        recommended : {false: "68", true: "96"}
    }
    const product = {
        name: "Camo Onesie"
    }
    
    it('Should correctly render the review list', () => {
        const reviewsTest = renderer
        .create(<Reviews reviewList={reviews} metaData={metaData} product={product} sortBy={()=>{}} getReviews={()=>{}}/>)
        .toJSON();
        expect(reviewsTest).toMatchSnapshot()
    });

    const wrapper = mount(<Reviews reviewList={reviews} metaData={metaData} product={product} sortBy={()=>{}} getReviews={()=>{}}/>)
    it('Should respond to show more reviews', () => {
        wrapper.find('.show-less').at(0).simulate('click')
        expect(reviews).toEqual(reviews)
    })

});