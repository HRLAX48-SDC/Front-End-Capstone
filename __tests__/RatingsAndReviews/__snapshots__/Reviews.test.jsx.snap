// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Reviews /> Should correctly render the review list 1`] = `
Array [
  <div
    className="card"
  >
    <div
      className="reviews-dropdown-stack hstack gap-2"
    >
      <div>
        Displaying
      </div>
      <b>
        1
         / 
        1
      </b>
      <div>
        reviews, sorted by:
      </div>
      <div
        className="dropdown"
      >
        <button
          aria-expanded={false}
          className="dropdown-toggle btn btn-outline-secondary"
          disabled={false}
          id="dropdown-basic-button"
          onClick={[Function]}
          type="button"
        >
          relevant
        </button>
      </div>
      <div
        className="ms-auto"
      >
        <div>
          <button
            className="open-modal btn btn-outline-secondary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Submit Review
          </button>
        </div>
      </div>
    </div>
    <div
      className="search-bar"
    >
      <form
        className=""
      >
        <div
          className="input-group"
        >
          <span
            className="input-group-text"
          >
            <i
              className="fas fa-search"
            />
          </span>
          <input
            className="form-control"
            onChange={[Function]}
            placeholder="Search reviews..."
            type="text"
            value=""
          />
        </div>
      </form>
    </div>
  </div>,
  <div
    className="reviews-list card"
  >
    <div
      className="card-body"
    >
      <div
        className="vstack gap-3"
      >
        <div
          className="card"
        >
          <div
            className="card-header"
          >
            <div
              className="hstack gap-3"
            >
              <span
                style={
                  Object {
                    "direction": "ltr",
                    "display": "inline-block",
                    "touchAction": "none",
                  }
                }
              >
                <span
                  aria-hidden="true"
                  className="react-simple-star-rating"
                  style={
                    Object {
                      "cursor": "",
                      "display": "inline-block",
                      "overflow": "hidden",
                      "position": "relative",
                      "userSelect": "none",
                      "verticalAlign": "middle",
                      "whiteSpace": "nowrap",
                    }
                  }
                >
                  <span
                    className="empty-icons"
                    style={
                      Object {
                        "color": "#cccccc",
                        "display": "inline-block",
                      }
                    }
                  >
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                  </span>
                  <span
                    className="filled-icons"
                    style={
                      Object {
                        "color": "#ffbc0b",
                        "display": "inline-block",
                        "left": 0,
                        "overflow": "hidden",
                        "position": "absolute",
                        "top": 0,
                        "transition": "",
                        "whiteSpace": "nowrap",
                        "width": "100%",
                      }
                    }
                    title="100 out of 5"
                  >
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                    <svg
                      className="star-svg"
                      fill="currentColor"
                      height={25}
                      style={Object {}}
                      viewBox="0 0 24 24"
                      width={25}
                    >
                      <path
                        d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                        fill="currentColor"
                        stroke="none"
                        strokeMiterlimit="10"
                        strokeWidth={0}
                      />
                    </svg>
                  </span>
                </span>
              </span>
              <div
                className="ms-auto"
              >
                BananaBoy123
              </div>
              <div
                className="review-tile-date"
              >
                November 9th, 2021
              </div>
            </div>
          </div>
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            />
            <p
              className="card-text"
            >
              Is this really the best purchase ever?
            </p>
            <div
              className="card"
            >
              <div
                className="card-body"
              >
                <p
                  className="card-text"
                >
                  Wow this product is amazing, 100/10 recommend. I own 1000 of these muahahaha. 
                </p>
                <div
                  className="row"
                />
              </div>
            </div>
            <div
              className="hstack gap-2"
            >
              <div>
                Was this review helpful?
              </div>
              <div
                className="review-helpful"
                onClick={[Function]}
              >
                Yes
              </div>
              <div>
                (
                6
                )
              </div>
              <div
                className="review-report"
                onClick={[Function]}
              >
                Report
              </div>
              <div
                className="ms-auto"
              >
                I recommend this product. ✔
              </div>
            </div>
          </div>
        </div>
        <button
          className="show-less btn btn-outline-secondary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Collapse
        </button>
      </div>
    </div>
  </div>,
]
`;
